prov/efa/src/efa_domain.c

```diff
@@ -88,18 +88,28 @@ static struct fi_ops_domain efa_ops_domain_rdm = {
  *
  * @param efa_domain[in,out]	efa domain to be set.
  * @param domain_name		domain name
+ * @param ep_type		endpoint type
  * @return 0 if efa_domain->device and efa_domain->ibv_pd has been set successfully
  *         negative error code if err is encountered
  */
-static int efa_domain_init_device_and_pd(struct efa_domain *efa_domain, const char *domain_name)
+static int efa_domain_init_device_and_pd(struct efa_domain *efa_domain,
+                                         const char *domain_name,
+                                         enum fi_ep_type ep_type)
 {
 	int i;
+	char *device_name = NULL;
+	const char *domain_name_suffix = efa_domain_name_suffix(ep_type);
 
 	if (!domain_name)
 		return -FI_EINVAL;
 
 	for (i = 0; i < g_device_cnt; i++) {
-		if (strstr(domain_name, g_device_list[i].ibv_ctx->device->name) == domain_name) {
+		device_name = g_device_list[i].ibv_ctx->device->name;
+		if (strstr(domain_name, device_name) == domain_name &&
+		    strlen(domain_name) - strlen(device_name) ==
+		            strlen(domain_name_suffix) &&
+		    strcmp((const char *) (domain_name + strlen(device_name)),
+		           domain_name_suffix) == 0) {
```

Comment:
The conditions inside the actual check here are a little wild. This is my suggestion:

Loop setup:

```c
// Get pointer to actual domain name suffix
const char *domain_name_suffix = memchr(domain_name, efa_domain_name_suffix(ep_type)[0], strlen(domain_name));
ptrdiff_t n_bytes = (ptrdiff_t) (domain_name_suffix - domain_name);
```

Actual check inside loop:

```c
// Short-circuits, avoids out-of-bounds access if device_name is too short
if (strlen(device_name) == n_bytes && memcmp(domain_name, device_name, n_bytes) == 0) {
  // ...
}
```

Reply:
Overall I think this approach adds more complexity without real benefits. Especially `memchr` appears unnecessary since we want to do full string match.

The if block looks larger than it is because of the line length restrictions. It looks reasonable if we increase the line length:
```c
if (strstr(domain_name, device_name) == domain_name &&
    strlen(domain_name) - strlen(device_name) == strlen(domain_name_suffix) &&
    strcmp((const char *) (domain_name + strlen(device_name)), domain_name_suffix) == 0)
```

Reply:
I am ok with the additional `strcmp((const char *) (domain_name + strlen(device_name)), domain_name_suffix) == 0`. This should make sure domain_name is a concatenation of device_name and domain_name_suffix